core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -242,7 +247,18 @@ private BytesToBytesMapIterator(
       }
     }
 
+    public BytesToBytesMapIterator setDestructive(boolean dt, BytesToBytesMap map) {
+      this.destructive = dt;
+      this.map = map;
+      return this;
+    }
+
     private void advanceToNextPage() {
+      if (destructive) {
+        dataPagesIterator.remove();
+        this.map.taskMemoryManager.freePage(currentPage);
+        this.map.shuffleMemoryManager.release(currentPage.size());
+      }
```

Comment:
@JoshRosen When we advance to new page, we free previous page here. I think this is what you meant?